feat(OAuth2): implement support for OAuth2 (1)

fioan89 opened the pull request:
Recent versions of Coder act as an OAuth 2.1 authorization server for first- and third‑party applications.
This PR aims at providing support for authenticating via OAuth with Coder Toolbox and still retain backward compatibility for authentication via API tokens or via certificates.

- implements #195 
- implements #267

This PR is a WIP:
- [x] Infrastructure code to discover if a Coder deployment supports OAuth2
- [x] Infrastructure code to discover the Coder OAuth2 endpoints
- [x] Infrastructure code to dynamically register Coder Toolbox as a client app
- [x] Toolbox API is configured so that authorization page can be launched
- [x] Token refresh is configured as well 
- [x] The infrastructure for exchanging the authorization codes for tokens and handle of auth redirect URI
- [x] Add support for PKCE verification
- [ ] Limit the permission scope to read workspaces, read templates, template versions, buildinfo, read/write workspace, agents + apps, read your own user details, authentication, read access to SSH keys
- [ ] Retrieve the logged account 
- [ ] Support for logout to revoke access/all tokens for the client app
- [x] Support for storing the oauth token and refresh token, re-use at the next start.
- [x] Update/Redesign the Login wizard screen to skip over the token step if OAuth is available
- [x] Redesigned the Settings page and regrouped all settings under a couple of logical topics
- [x] Integrate with Coder CLI, authenticate the CLI with OAuth2 as well.
- [ ]  Update documentation

fioan89 committed:
build: simplify install folder resolution

fioan89 committed:
impl: setup auth manager with auth and token endpoints

Toolbox API comes with a basic oauth2 client. This commit
sets-up details about two important oauth flows:

- authorization flow, in which the user is sent to web page
  where an authorization code is generated which is exchanged
  for an access token.
- details about token refresh endpoint where users can obtain
  a new access token and a new refresh token.

A couple of important aspects:
- the client app id is resolved in upstream
- as well as the actual endpoints for authorization and token refresh
- S256 is the only code challenge supported

fioan89 committed:
impl: retrieve supported response type and the dynamic client registration url

OAuth endpoint `.well-known/oauth-authorization-server` provides metadata about
the endpoint for dynamic client registration and supported response types.
This commit adds support for deserializing these values.

fioan89 committed:
impl: models for dynamic client registration

OAuth allows programatic client registration for apps like Coder Toolbox
via the DCR endpoint which requires a name for the client app, the requested
scopes, redirect URI, etc... DCR replies back with a similar structure but
in addition it returs two very important properties: client_id - a unique
client identifier string and also a client_secret - a secret string value
used by clients to authenticate to the token endpoint.

fioan89 committed:
impl: pixy secure code generator

Code Toolbox plugin should protect against authorization code interception
attacks by making use of the PKCE security extension which involves
a cryptographically random string (128 characters) known as code verifier
and a code challenge - derived from code verifier using the S256 challenge method.

fioan89 committed:
impl: retrofit API for endpoint discovery and dynamic client registration

fioan89 committed:
impl: factory method for the auth manager

The OAuth2-compatible authentication manager provided by Toolbox

fioan89 committed:
impl: improve auth manager config

- authentication and token endpoints are now passed via the login configuration object
- similar for client_id and client_secret
- PCKE is now enabled

fioan89 committed:
refactor: simplify OAuth manager architecture and improve dependency injection

- remove ServiceLocator dependency from CoderToolboxContext
- move OAuth manager creation to CoderToolboxExtension for cleaner separation
- Refactor CoderOAuthManager to use configuration-based approach instead of constructor injection

The idea behind these changes is that createRefreshConfig API does not receive a configuration
object that can provide the client id and secret and even the refresh url. So initially
we worked around the issue by passing the necessary data via the constructor. However this approach
means a couple of things:

- the actual auth manager can be created only at a very late stage, when a URL is provided by users
- can't easily pass arround the auth manager without coupling the components
- have to recreate a new auth manager instance if the user logs out and logs in to a different URL
- service locator needs to be passed around because this is the actual factory of oauth managers in Toolbox

Instead, we went with a differet approach, COderOAuthManager will derive and store the refresh configs once
the authorization config is received. If the user logs out and logs in to a different URL the refresh data is
also guaranteed to be updated. And on top of that - this approach allows us to get rid of all of the issues
mentioned above.

fioan89 committed:
fix: inject mocked PluginAuthManager into UTs

fioan89 committed:
impl: handle the redirect URI

Toolbox can handle automatically the exchange of an authorization code with a token
by handling the custom URI for oauth. This commit calls the necessary API
in the Coder Toolbox URI handling.

fioan89 committed:
fix: wrong client app registration endpoint

POST /api/v2/oauth2-provider/apps is actually for manual admin
registration for admin created apps. Programmatic Dynamic Client
Registration is done via `POST /oauth2/register`.

At the same time I included `registration_access_token` and `registration_client_uri`
to use it later in order to refresh the client secret without re-registering the client app.

fioan89 committed:
impl: simple way of triggering the OAuth flow.

A bunch of code thrown around to launch the OAuth flow.
Still needs a couple of things:
- persist the client id and registration uri and token
- re-use client id instead of re-register every time
- properly handle scenarios where OAuth is not available
- the OAuth right now can be enabled if we log out and then
hit next in the deployment screen

fioan89 committed:
Merge branch 'main' into impl-support-for-oauth

fioan89 committed:
Merge branch 'main' into impl-support-for-oauth

fioan89 committed:
impl: add config to enforce auth via API token

A new config `preferAuthViaApiToken` allows users to continue to use
API tokens for authentication when OAuth2 is available on the Coder deployment.

fioan89 committed:
Merge branch 'main' into impl-support-for-oauth

fioan89 committed:
Merge branch 'main' into impl-support-for-oauth

fioan89 committed:
impl: resolve the account

Account implementation with logic to resolve the account once the token
is retrieved. Marshalling logic for the account is also added.

There is a limitation in the Toolbox API where createRefreshConfig
is not receiving the auth params. We worked around by capturing and
storing these params in the createAuthConfig but this is unreliable.
Instead we use the account to pass the missing info around.

fioan89 committed:
chore: fix UTs

fioan89 committed:
impl: rework the first login screen and discover if oauth2 is supported

OAuth2 should be launched if user prefers is over any other method of auth
and if only the server supports it.

fioan89 committed:
impl: prefer client_secret_post as token auth method if available

Fallback on client_secret_basic or None depending on what the Coder
server supports.

fioan89 committed:
fix: missing client secret from authorization request

fioan89 committed:
fix: prefer client_secret_basic auth method

As per https://github.com/coder/coder/blob/main/docs/admin/integrations/oauth2-provider.md#client-authentication-methods

fioan89 committed:
impl: support for client_secret_basic and client_secret_post for token endpoint

Based on the auth method type we need to send client id and client secret as a basic
auth header or part of the body as an encoded url form

fioan89 committed:
impl: implement our own OAuth2 client (1)

We encountered a couple of issues with the Toolbox API which is inflexible:
- we don't have complete control over which parameters are sent as query&body
- we don't have fully basic + headers + body logging for debugging purposes
- doesn't integrate that well with our existing http client used for polling
- spent more than a couple of hours trying to understand why Coder rejects the
  authorization call with:
 ```
  {"error":"invalid_request","error_description":"The request is missing required parameters or is otherwise malformed"} from Coder server.
 ```
 Instead we will slowly discard the existing logic and rely on enhancements to our existing http client.
 Basically, the login screen will try to first determine if mTLS auth is configured and use that, otherwise
 it will check if the user wants to use OAuth over API token, if available. When the flag is
 true then the login screen will query the Coder server to see if OAuth2 is supported.
 If that is true then browser is launched pointing to the authentication URL. If not we will default to
 the API token authentication.

fioan89 committed:
impl: implement our own OAuth2 client (2)

Same pattern for instantiating the same http client is used in many places.
Refactored the setup into a simple factory method.

fioan89 committed:
impl: implement our own OAuth2 client (3)

Implements the callback for the authentication step. It basically exchanges the auth code for
a token response. Supports client_secret_basic as preferred auth method by Coder for the
token endpoint, but also client_secret_post if nothing is available.

fioan89 committed:
fix: code challenge was sent twice to the auth endpoint

Coder server effectively rejected the request because the
challenge was executed twice

fioan89 committed:
fix: include state for cross-checking later when the auth code is returned

fioan89 committed:
fix: short circuit the URI handler when handling oauth callbacks

fioan89 committed:
fix: short circuit the URI handler when handling oauth callbacks (2)

Include utility for fallbacks of auth method.

fioan89 committed:
impl: implement our own OAuth2 client (4)

Initiates the REST API client and the CLI with the resolved token.
There is no support for refreshing the token if it expires, for now.

fioan89 committed:
impl: support for token refresh (1)

Rough implementation for detecting expired tokens, and calling the token endpoint
to refresh and retrieve a new access token but also a new refresh token.

If the refresh is successful then we retry the failed call again, but also
execute a callback that in subsequent commits should allow the new token to be
passed to the cli, and also stored.

This is a crude implementation, I think a refactor is needed as there are so many params
that are cloned and passed around between the CoderProvider, login steps and rest api client
and then back again to the coder remote provider.

fioan89 committed:
chore: rename context class

Which was improperly named, it had nothing to do with the CLI.
It is used only for the auth setup wizard.

fioan89 committed:
fix: broken UTs

fioan89 committed:
chore: dependency declared twice in the build system

fioan89 committed:
fix: dependingon the JDK vendor and version zt-exec can raise IOException instead of ProcessInitException

We observed some UTs fails to a difference in the JDK version used by Gradle (CLI) versus the one used by IntelliJ to run the tests.
The zt-exec library attempts to wrap the standard java.io.IOException (thrown when a process fails to start) into a ProcessInitException.
It does this by parsing the exception message to extract the error code (e.g., looking for error=2).
However, different JDK versions (and distributions) format this error message differently.
If the JDK used by IntelliJ produces an error message that zt-exec doesn't recognize (e.g., missing the error= part or using a different format),
zt-exec fails to wrap the exception and instead propagates the original IOException.
To fix the test so it passes in both environments, we should catch the more general IOException (which is the parent of ProcessInitException) and
assert on the essential part of the error message ("No such file or directory") rather than the specific format zt-exec produces.

fioan89 committed:
impl: persist refresh token between Toolbox restarts (1)

We need the client id, client secret, auth method a refresh token to be persisted
so that at the next app restart we can avoid going again through client registration,
and authentication and authorization steps.

fioan89 committed:
impl: persist refresh token between Toolbox restarts (2)

Save the oauth session details as soon as the http client and cli
are initialized. Up until know it happened only on token refresh.

fioan89 committed:
impl: persist refresh token between Toolbox restarts (3)

Load previous oauth session details, and resolve the access token
before initializing the cli and rest client.

fioan89 committed:
impl: support logging out and logging back in via OAuth2

A flow like logging out from an existing OAuth2 session and then logging
back in fails due to issues with handleUri API which is not able to properly
handle UI pages, especially if an existing page was already pushed into the
display stack - after log out we have the DeploymentUrl step visible. Once the user
hits next the browser is launched where the user authorizes the client app. After
which the callback URL pops up the plugin screen which is still showing the deployment
url. Normally, in the URI handler we would now push a new page - the ConnectStep. But
that doesn't work, the plugin is still stuck with the initial page. After extensive
research I found way around by making the login wizard for "listening" to step updates.

In other words, once the authorization callback launches the URI, in the handler instead
of pushing a new page to the display stack we actually push a new "step" in the existing
wizard page. See https://youtrack.jetbrains.com/issue/TBX-16622 for more
details.

fioan89 committed:
fix: properly close the setup wizard if connection is successful

There were a couple of issues with the existing implementation.
ConnectStep was emiting the URL_REQUEST both at the end of the connect
but also part of the close() call. This caused the screen to flicker
during setup. As a fix close no longer resets the setup global state
while a final new step was introduced to allow the ConnectStep to
properly close the entire wizard.

fioan89 committed:
fix: mark the header bar as busy

When switching between Coder deployments using the URI.

fioan89 committed:
Merge branch 'main' into impl-support-for-oauth

fioan89 committed:
Merge branch 'main' into impl-support-for-oauth

fioan89 committed:
impl: new setting that allows user to use OAuth over API token authorization

mTLS always takes precedence over OAuth and API token authorization, while API Token
is the default method if mTLS is not used. But now user can prefer OAuth authorization
over API tokens, if OAuth is available. The setting is now available in the Settings page,
no longer in the main login screen as it was the case in the initial draft of the PR.

fioan89 committed:
impl: group the settings in a couple of sections

For easy navigation as there were too many of them, mixed
with no logical differentiation. Now we cluster on a couple of topics:
- General
- Security & Authentication
- CLI
- SSH

fioan89 committed:
chore: remove unused data models

Initially the OAuth implementation was trying to use
the existing OAuth API/Client provided by Toolbox which is not very flexible.

jeremyruppel commented on src/main/kotlin/com/coder/toolbox/store/CoderSecretsStore.kt:
```diff
@@ -14,9 +22,38 @@ class CoderSecretsStore(private val store: PluginSecretStore) {
+        )
+    }
+
+    fun storeOAuthFor(url: String, oAuthSessionCtx: CoderOAuthSessionContext) {
```
Nvm, you need the full deployment url as the key, so you'd just be `toString()`ing it here anyways 👍

fioan89 committed:
fix: handle invalid state responses

The OAuth2 server implementation needs to provide an authorization code
that can be exchanged for an access token. But in order to make sure the
authorization code is for the "our" login request, the client provides a
state value when launching the authorization URL which the OAuth2 server has
to send back when with the auth code. This fix makes sure the authorization
code is actually sent, and that the state value is the same as in our initial
request.

fioan89 committed:
fix: report errors when token exchange fails

This fix reports an error to the user when token exchange
request is failing, or returning an empty body or a body that does not contain the token.

fioan89 committed:
chore: remove non nullable assertions

fioan89 committed:
chore: refactor and reuse OAuth2 logic

The logic for exchanging auth code to tokens, refreshing tokens was used in
multiple places without any code reuse strategy. Extracted an OAuth service that
handles the basic operations.

fioan89 committed:
fix: use the registration url from the metadata

The metadata endpoint provide an absolute URL for the client registration endpoint
which we should use instead of hardcoding the path relative to the base url.

fioan89 committed:
impl: handle client registration errors

https://datatracker.ietf.org/doc/html/rfc7591 normalizes the client
registration error responses and forces providers to always include
json with an error code and an error message. This patch captures the error
response and builds a pretty message and displays it to the user.

fioan89 committed:
impl: handle auth redirect and token registration errors

RFC 6749 §4.1.2.1 + RFC 7636 §4.4.1 specify that the error
code and optional error_description can be returned as a query params
int the callback URI.

Similarly, RFC 6749 §5.2 — the exchange of authorization codes to tokens
can return a json body containing an error code and an error message that
was never handled in our code.

fioan89 committed:
chore: update scopes

fioan89 committed:
build: upgrade TBX API to version 1.10.76281

This upgrade will need TBX 3.4 or higher to be installed.
The upgrade is needed to benefit from the fixes related to displaying UI
pages in the URI handler.

In addition I reworked the main build.gradle and extracted everything into a
small custom plugin.

fioan89 committed:
chore: next version is 0.9.0

Due to the dependency on the new API.

fioan89 committed:
fix: properly decode query params

OAuth callbacks are encoded, especially error details
need to be decoded before surfacing them to the user.

fioan89 committed:
fix: use - as separator for error codes and error descriptions

We ended up with error messages like `An error was encountered: <error-code>: <some error description`. ":" is a bit repetitive.

fioan89 committed:
Merge branch 'main' into impl-support-for-oauth

matifali reviewed (approved):
Non-engineering approval. I am fine shipping this, given its opt-in.<br>Just ensure we provide a good experience in case the setting is enabled, but the deployment does not have `CODER_EXPERIMENTS=aouth2` enabled.

fioan89 commented:
> Non-engineering approval. I am fine shipping this, given its opt-in.Just ensure we provide a good experience in case the setting is enabled, but the deployment does not have `CODER_EXPERIMENTS=aouth2` enabled.

Yes, I can confirm that OAuth happens only when the user explicitly enables the OAuth authentication AND the backed exposes the necessary endpoints.

code-asher commented on src/main/kotlin/com/coder/toolbox/oauth/OAuth2Service.kt:
```diff
@@ -0,0 +1,177 @@
+            else -> ClientAuth.None(oauthSessionContext.clientId)
+        }
+
+        val service = createAuthorizationService()
```
Is there a reason we need to create a new service here rather than use the one on the class?  Or should we update the service on the class with this new one?

fioan89 replied:
No it just shadows the class level one. Unfortunately I implemented this over a span of a couple of months with lots of context switches on other priorities.  So there is a high chance stuff like this happened in multiple places. :(

code-asher commented on src/main/kotlin/com/coder/toolbox/oauth/OAuth2Client.kt:
```diff
@@ -0,0 +1,177 @@
+        throw Exception(errorMessage)
+    }
+
+    private fun createAuthorizationService(): CoderAuthorizationApi {
```
This is a nit, but when I initially saw the oauth2 auth service create yet another auth service it seemed weird to me, but this is really just an http/api client right?  Not really doing any service-like things, I think?  

In my mind it would involve state management to be a service, which has implications for how it should behave in the code (which is why I thought maybe it was a problem to recreate it above without updating the one on the class).

I guess in that sense `OAuth2Service` is not necessarily a service either, just a wrapper around the API calls.  Actually, could these all be methods directly on the coder rest client?  Feels to me like it could be part of the sdk, they are just more API endpoints after all.

fioan89 replied:
Hmm... this is an interesting point and I'd like to discuss/philosophize a bit over it.
First - CoderAuthorizationApi is just a Retrofit interface — it's an HTTP client definition, OAuth2Service is stateless — it just wraps those API calls with error handling.                                                                                                            

Could these live on CoderRestClient? In principle yes — they're just more HTTP calls to the Coder deployment. But there's a practical issue : OAuth2Service is used before CoderRestClient exists.                                                       The discovery/registration/exchange calls are pre-authentication — they use a bare HTTP client (no auth interceptors, no token). CoderRestClient is constructed with a token or OAuth context already in hand, and its  HTTP client is configured with auth interceptors. Mixing unauthenticated OAuth endpoint calls into that client would mean either:                                                                                      
                                                                                                                                                                                                                         
1. Building a second internal HTTP client without auth interceptors
2. Making the auth interceptors conditional per-request
                                                                                                                                                                                                                         
Both add complexity for little gain.  So I think the current separation makes sense architecturally 

Now regarding the naming - services usually orchestrate business logic. I guess I confused request construction with logic. I'll try to come up with a better name.

code-asher replied:
Oh yeah true, good points.

code-asher commented on src/main/kotlin/com/coder/toolbox/sdk/CoderRestClient.kt:
```diff
@@ -356,6 +387,13 @@ open class CoderRestClient(
 
+    private suspend fun refreshToken() {
+        val newAuthResponse = OAuth2Service(context).refreshToken(oauthContext!!)
+        this.oauthContext.tokenResponse = newAuthResponse
```
nbd at all but we use `oauthContext` without a `this` and then `this.oauthContext`, is there a reason for that?

I know I always say this haha but `!!` feels like a trap waiting to spring in the future, maybe we could pass in the context to `refreshToken()` or something?

fioan89 replied:
Arghhhh....I always tell myself - let's quickly throw two ! together and I'll rewrite the code later, for now let's make it work. As you have noticed...that later never comes :)

code-asher replied:
relatable lol

code-asher commented on src/main/kotlin/com/coder/toolbox/sdk/CoderRestClient.kt:
```diff
@@ -342,8 +352,29 @@ open class CoderRestClient(
-            block()
+        try {
+            val response = block()
+            if (response.code() == HttpURLConnection.HTTP_UNAUTHORIZED && oauthContext.hasRefreshToken()) {
```
If `oauthContext` is nullable would this not require a `?`?  Should we do an `oauthContext.let` or something?  Would also let us get rid of that `!!` if we pass that around.

But I assume it must not require it since it is building, just not sure how haha

fioan89 replied:
It works because hasRefreshToken is an extension function on a nullable context:

```
fun CoderOAuthSessionContext?.hasRefreshToken(): Boolean = this?.tokenResponse?.refreshToken != null
```

code-asher replied:
Ooooooo did not know that was possible.  Neat.

code-asher commented on src/main/kotlin/com/coder/toolbox/sdk/CoderRestClient.kt:
```diff
@@ -342,8 +356,29 @@ open class CoderRestClient(
-            block()
+        try {
+            val response = block()
+            if (response.code() == HttpURLConnection.HTTP_UNAUTHORIZED && oauthContext.hasRefreshToken()) {
```
Is there a downside to moving this call inside the mutex?  It does feel a little sus, even if currently there is not a code path that triggers a race.

code-asher commented on src/main/kotlin/com/coder/toolbox/sdk/CoderRestClient.kt:
```diff
@@ -356,6 +387,13 @@ open class CoderRestClient(
+    private suspend fun refreshToken() {
+        val newAuthResponse = OAuth2Service(context).refreshToken(oauthContext!!)
+        this.oauthContext.tokenResponse = newAuthResponse
+        onTokenRefreshed?.invoke(url, oauthContext)
```
Probably not a big deal at all, but it could be kinda weird that `onTokenRefreshed` could technically mutate the context.  `refreshToken` could too, technically.

fioan89 replied:
Done.

code-asher commented on src/main/kotlin/com/coder/toolbox/sdk/CoderRestClient.kt:
```diff
@@ -342,8 +352,29 @@ open class CoderRestClient(
+                        refreshToken()
+                        true
+                    } catch (e: Exception) {
+                        context.logger.error(e, "Failed to refresh access token")
```
If the refresh fails, do we need to look at the response and possibly discard the token?  If it is some kind of permanent auth failure.  Otherwise I imagine we would keep trying to refresh with the same token.

fioan89 replied:
I think I'll log an issue. There are transient reasons (network failure??) for which it is worth retrying at the next API call. But for other errors like invalid_grant or unauthorized_client it is probably pointless to retry again. And it is not easy to tackle this problem - do we disrupt the user and remove his workspaces, stop the ongoing ssh connections and popup the login screen, or do we stop the world but still keep the workspaces visible? I think it is worth pondering over and maybe involve @matifali in the discussion, but in a separate ticket.

fioan89 replied:
https://github.com/coder/coder-jetbrains-toolbox/issues/273

code-asher reviewed (commented):
I ran out of time but will finish tomorrow!

code-asher commented on src/main/kotlin/com/coder/toolbox/views/ConnectStep.kt:
```diff
@@ -87,12 +95,22 @@ class ConnectStep(
             try {
+                var oauthSession: CoderOAuthSessionContext? = null
+                if (context.settingsStore.requiresTokenAuth && context.settingsStore.preferOAuth2IfAvailable && CoderSetupWizardContext.hasOAuthSession()) {
+                    refreshOAuthToken()
+                    oauthSession = CoderSetupWizardContext.oauthSession!!.copy()
```
Could we call refreshToken on the rest client instead of duplicating it here?  Would have to create the rest client first of course.  Mostly it feels kinda scattered to me with how we have two copies of oauth session context and we have to kinda glue them together.

Also maybe we could return the session to avoid `!!`.

fioan89 replied:
I tried to do it quickly - but there are a number of issues:

- cli also needs to be initialized
- I need to find a good way to do the refresh token in the http client before any other API call is made.
Is it alright if I log an issue and treat this separetly?

fioan89 replied:
>  if I deselect the oauth option in the settings it seems to immediately go back to using the API key. But the reverse seems not to be true

This sounds really scary. If you are logged via oauth, and go to the Settings page and deselect the "Prefer OAuth.." the API token should be used only at the next restart or if you log out and log in again. The client should not terminate automatically, and go through the login screen again. Is this what you are experiencing (I can't reproduce it)

And yeah... it was intentional to keep the API token stored.

@code-asher do you think I can treat all of these into a separate issue/PR? it is a lot of stuff that I need to test and think about it. I will need Atif's help to decide the behavior for some of these scenarios (like removing the API token)

code-asher replied:
> do you think I can treat all of these into a separate issue/PR

Yes of course!

> If you are logged via oauth, and go to the Settings page and deselect the "Prefer OAuth.." the API token should be used only at the next restart or if you log out and log in again. The client should not terminate automatically, and go through the login screen again. Is this what you are experiencing

What happened is that I deselected "prefer oauth", then I revoked permissions for the oauth app in my Coder dashboard.  But it kept connecting to workspaces just fine.

To me, this felt unexpected, I thought the current session would keep using oauth (and so connecting would result in an auth failure), and for a second it felt like somehow it had bypassed auth lol until I realized it was probably just using my old API key I had configured before switching to oauth.

I think it is because we check the "prefer oauth" setting each time rather than having that info embedded as part of the session itself.  IMO it should only affect the decision we make when first logging in.

fioan89 replied:
Raised the following issues:

https://linear.app/codercom/issue/DEVEX-221/unexpected-behavior-when-user-deselects-the-prefer-auth-setting-in
https://linear.app/codercom/issue/DEVEX-222/misleading-error-after-disabling-oauth-in-coder-toolbox
https://linear.app/codercom/issue/DEVEX-223/setup-is-ongoing-even-after-user-hit-back-in-the-setup-screen-for

code-asher reviewed (commented):
Tried it out and oauth worked fabulously!  Just FYI I only reviewed the oauth stuff and skipped `buildSrc`.

Had some observations (some of these might be pre-existing issues):

1. If I deselect the oauth option in the settings it seems to immediately go back to using the API key.  But the reverse seems not to be true, if I select oauth while already logged in then it keeps using my API key (until I restart Toolbox).

Is that what we want?  I think I expected my session to keep using whatever it used to log in until I explicitly logged out.  I also kind of expected my API key to be deleted once I switched to oauth (or rather I expected it would be overridden with the oauth access token), but it seems to have been preserved, so even if I explicitly log out and then disable oauth, it still authenticates with the API key which was surprising to me.

2. This one is kinda wacky and probably not going to happen in practice so might not be worth addressing, but if I try to log in, then disable oauth, then click "allow", the error message does not really make sense, says "oauth or api token is required" but I feel like the error should really be something like "unable to log in with oauth because it is disabled".  Or, we should remember what we used to start the login and carry that through the whole process.

3. When I launched Toolbox I got "Error encountered while setting up Coder" and "authorization failed" which looks a bit scary but according to the logs my API key was invalid (makes sense, I have not launched Toolbox in a long while).  We should add the actual error message to that dialog.

4. If I click out of Toolbox while the cli is downloading (for example if I close the browser window opened by the oauth process) then Toolbox closes itself (idk why they made it like this) and if I re-open it I am back on the URL screen.  It looks like in the background the setup is still ongoing though.  If I try to log in again, they seem to clash and I do get the workspaces list but I also get the security dialog (maybe because it is trying to use a file that got overwritten by the other).  Same thing happens if I click "back" and log in again.  Not sure what happens if I do that and try to log into a different deployment while the other is ongoing.

fioan89 committed:
chore: refactor shadowing service

fioan89 committed:
chore: rename OAuth2 http client

Currently, named OAuth2Service which oversells itself because it
only orchestrates http request construction, there is no business
logic orchestration or state management.

fioan89 committed:
chore: get rid of !! from refreshToken implementation

fioan89 committed:
chore: refactor oauth context mutation

We used to share oauth context as a global val that could
be mutated once a token had to refreshed. Instead, we changed
the code pass a modified copy of this oauth context with the
refreshed token.

code-asher reviewed (approved):
Putting down an approval, we can follow up on anything remaining separately!
